sound/soc/sof/ipc4-topology.c

```diff
@@ -1222,6 +1222,68 @@ static int snd_sof_get_nhlt_endpoint_data(struct snd_sof_dev *sdev, struct snd_s
 }
 #endif
 
+static int ipc4_set_fmt_mask(struct snd_mask *fmt, unsigned int bit_depth)
+{
+	switch (bit_depth) {
+	case 16:
+		snd_mask_set_format(fmt, SNDRV_PCM_FORMAT_S16_LE);
+		break;
+	case 24:
+		snd_mask_set_format(fmt, SNDRV_PCM_FORMAT_S24_LE);
+		break;
+	case 32:
+		snd_mask_set_format(fmt, SNDRV_PCM_FORMAT_S32_LE);
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int ipc4_copier_set_capture_fmt(struct snd_pcm_hw_params *pipeline_params,
+				       struct snd_pcm_hw_params *fe_params,
+				       struct sof_ipc4_available_audio_format *available_fmt)
+{
+	struct sof_ipc4_audio_format *audio_fmt;
+	unsigned int sample_valid_bits;
+	bool multiple_formats = false;
+	bool fe_format_match = false;
+	struct snd_mask *fmt;
+	int i;
+
+	for (i = 0; i < available_fmt->num_output_formats; i++) {
+		unsigned int val;
+
+		audio_fmt = &available_fmt->output_pin_fmts[i].audio_fmt;
+		val = SOF_IPC4_AUDIO_FORMAT_CFG_V_BIT_DEPTH(audio_fmt->fmt_cfg);
+
+		if (i == 0)
+			sample_valid_bits = val;
+		else if (sample_valid_bits != val)
+			multiple_formats = true;
+
+		if (snd_pcm_format_width(params_format(fe_params)) == val)
+			fe_format_match = true;
+	}
+
+	/* case 1: topology not providing enough info */
+	if (multiple_formats && !fe_format_match)
+		return -EINVAL;
+
+	fmt = hw_param_mask(pipeline_params, SNDRV_PCM_HW_PARAM_FORMAT);
+	snd_mask_none(fmt);
+
+	/* case 2: multiple formats defined and one matches FE */
+	if (multiple_formats) {
+		snd_mask_set_format(fmt, params_format(fe_params));
+		return 0;
+	}
+
+	/* case 3: single format defined in topology, use that */
+	return ipc4_set_fmt_mask(fmt, sample_valid_bits);
+}
```

Comment:
case 4: single format defined in topology and it is not matching?

Reply:
@ujfalusi that's case 3, if there's just a single format, we use that whether it matches with FE or not. the pipeline is guaranteed to have another copier that can do the conversion, so this should be always good. topology developer wanted to use this format, so let's use what is requested.

case-1 is a bit different as kernel needs to pick one, but has no context to make a good choice. should we pick the first one, or the format with highest precision...?